include/stdexec/execution.hpp

```diff
@@ -3103,76 +3096,123 @@ namespace stdexec {
+        // __notify function is called from the shared state's __notify_waiters function, which
+        // first sets __waiters_ to the completed state. As a result, the attempt to remove `this`
+        // from the waiters list above will fail and this stop request is ignored.
+        __release_shared_state();
+        stdexec::set_stopped(static_cast<_Receiver&&>(this->__receiver()));
+      }
+
+      // This is called from __shared_state::__notify_waiters when the input async operation
+      // completes; or, if it has already completed when start is called, it is called from start:
+      // __notify cannot race with __on_stop_request. See comment in __on_stop_request.
       template <class _Tag>
-      static void __action(__local_state_base* __self, __action_kind __kind) noexcept {
-        auto* const __op = static_cast<__local_state*>(__self);
-        if (__kind == __action_kind::__notify) {
-          __op->__on_stop_.reset();
-
-          // The split algorithm sends by T const&. ensure_started sends by T&&.
-          if constexpr (same_as<__split::__split_t, _Tag>) {
-            std::visit(
-              __notify_visitor(__op->__receiver()), std::as_const(__op->__shared_state_->__data_));
-          } else {
-            std::visit(
-              __notify_visitor(__op->__receiver()), std::move(__op->__shared_state_->__data_));
-          }
-        } else {
-          // This is a detach operation
-          if constexpr (same_as<__split::__split_t, _Tag>) {
-            // no-op
-          } else {
-            __op->__shared_state_->__detach();
-          }
+      static void __notify(__local_state_base* __base) noexcept {
+        auto* const __self = static_cast<__local_state*>(__base);
+
+        // The split algorithm sends by T const&. ensure_started sends by T&&.
+        constexpr bool __is_split = same_as<__split::__split_t, _Tag>;
+        using __variant_t = decltype(__self->__sh_state_->__results_);
+        using __cv_variant_t = __if_c<__is_split, const __variant_t&, __variant_t>;
+
+        __self->__on_stop_.reset();
+
+        // Copy this into a local because the call to std::visit can cause __self to be deleted.
+        auto* __sh_state = __self->__sh_state_;
+        bool const __holds_ref_ = __self->__holds_ref_.exchange(false);
+
+        auto __visitor = __make_notify_visitor(__self->__receiver());
+        std::visit(__visitor, static_cast<__cv_variant_t&&>(__sh_state->__results_));
+
+        // Release the reference on the shared state after we complete the receiver since it is
+        // possible for this to cause the destruction of the shared state. (See `start` below.)
+        if (__holds_ref_) {
+          __sh_state->__dec_ref();
         }
       }
 
       std::optional<__on_stop_cb_t> __on_stop_{};
-      __intrusive_ptr<__shared_state_t> __shared_state_;
+      __sh_state_t* __sh_state_;
+      std::atomic<bool> __holds_ref_{true};
```

Comment:
It's not clear why you need an atomic bool here.

The `__notify` and `__on_stop_request` functions should not race with regards to which will decrement the ref-count - if `__on_stop_request` decrements the ref-count then `__notify` will not be run.

And you shouldn't be racing between the destructor of the op-state being run and either of these function calls.

Can you just have this class hold one of the `__data<__sh_state_t>` values that manages the ref-count for you?

Reply:
I initially implemented this without the atomic bool but tsan cried foul.